src/core/or/policies.c

```diff
@@ -495,8 +500,39 @@ fascist_firewall_prefer_ipv6_impl(const or_options_t *options)
 MOCK_IMPL(int,
 fascist_firewall_rand_prefer_ipv6_addr, (void))
 {
-  /* TODO: Check for failures, and infer our preference based on this. */
-  return crypto_rand_int(2);
+  or_state_t *state = get_or_state();
+  time_t current_time = time(NULL);
+
+  /* Clear IPv4 and IPv6 failure entries if they are more than a week old. */
+  if (state->IPv4AutoFailTime &&
+      state->IPv4AutoFailTime <= current_time - IP_FAIL_CLEAR_TIME) {
+    state->IPv4AutoFail = 0;
+    state->IPv4AutoFailTime = 0;
+  }
+  if (state->IPv6AutoFailTime &&
+      state->IPv6AutoFailTime <= current_time - IP_FAIL_CLEAR_TIME) {
+    state->IPv6AutoFail = 0;
+    state->IPv6AutoFailTime = 0;
+  }
+
+  if (state->IPv4AutoFail && !state->IPv6AutoFail) {
+    /* If we fail IPv4 automatically, prefer IPv6. */
+    return 1;
+  } else if (!state->IPv4AutoFail && state->IPv6AutoFail) {
+    /* If we fail IPv6 automatically, prefer IPv4. */
+    return 0;
+  } else if (state->IPv4AutoFail && state->IPv6AutoFail) {
+    /* If we fail both IPv4 and IPv6, reset the auto fail counters and then
+     * select randomly. */
+    state->IPv4AutoFail = state->IPv6AutoFail = 0;
+  }
+
+  int fail_sum = state->IPv4Fails + state->IPv6Fails;
+  int chosen_number = crypto_rand_int(fail_sum);
```

Comment:
Asking for up to 32 bits of entropy for each new circuit could significantly slow down busy clients. I think that @nmathewson has made our PRNG fast enough to handle this kind of load.

But we might want to think about this problem differently.
Let's work out an acceptable failure rate for connections, for example, 1/8.
Then, we reduce the probability of IPv4/IPv6 to one of a small set of probabilities: 1/8, 2/8, 3/8, ... , 7/8.
(Deliberately excluding 0/8 and 8/8.)
This design is better, because it doesn't starve clients that have just changed networks.